### ISSUE TITLE
Add a Gaussian Blur Compute Shader extension to irr::ext

### ISSUE THREAD

**Issue by devshgraphicsprogramming** [buildaworldnet/IrrlichtBAW#70]

**Comment by devshgraphicsprogramming** [comment]
Done pending some improvements.

Bug fix:
Right and up overflow on presum is not handled correctly giving overbright areas in the X+and y+ edges

While bugfixing change the CBlurPerformer to take RADIUS as a UBO parameter (there is oneplace of useless padding where you can fit a 32bit float) and add a blur radius parameter to `CBlurPerformer::writeUBO` where the blur radius is number from 0 to 1 indicating  a % of the X and Y size of the texture.
Finally allow sampling of non-integer coordinates in the box blur from the prefix sum array by implementig a variant of linear interpolatin of the nearest two samples with mix(,fract()) at the front and end.

To change for final usability:
1) split `createBlurredTexture` into `createOutputTexture` (which will be optional to use) and `blurTexture`, user can provide their own texture it just has to use the rgba16f format (check with assert) and be larger or equal in width and height as the desired outpu texure.
2) Stop creating and dropping the texture every frame (but blur evey frame) in the example
3) Aloow the usage of user's own custom UBO (have writeUBO write to a void* ptr the data needed for a blur pass) and change example so it provides custom larger ubo
4) Allow the blurring of non-square textures
5) Merge the presum compute shader with the blur compute shader (the data doesn't need to leave shared memory, the blurred image just needs to be below 1024 pixels in size - or 2048 on non-mobile but lets take lowest common denominator)
6) Plan for the case when maximum work group size is 256 but blurred image size is >256
7) Choose initial downscaling (create a downscale compute shader on the fly) with configurable MxN downscale where M and N can be independently between 1 and 16 

For Max workgroup and shared memory  size related capability queries I can add these to the engine.

**Comment by devshgraphicsprogramming** [comment]
@Crisspl  did you manage to fix 

> Bug fix:
Right and up overflow on presum is not handled correctly giving overbright areas in the X+and y+ edges

?

**Comment by devshgraphicsprogramming** [comment]
nvm I see
https://github.com/Crisspl/IrrlichtBAW/commit/ead252d6331bab4e64ede754a7349e340eb69504
fixed it

**Comment by Crisspl** [comment]
I'll just paste it here:
1) roll the downsample into the horizontal pass shader
2) collapse all downsweeps and upsweeps  for the same axis into one shader
3) take away u_iterNum,u_outMlt,u_outOffset and u_inOffset and compile them in since we will only have 2 shaders
4) add an option at startup to CBlurPerformer of how many box blur passes to do (some people say 2 not 3 is enough)
5) make an option (template or something) to output to texture `GL_RG9_E5` (bind output image as `r32u`i)
6) ~~In `prepSmemForPresum `:~~
 ```
storeShared(_tid,decodeRGB());
if (_tid<PS_SIZE-ACTUAL_SIZE)
storeShared(_tid+ACTUAL_SIZE,vec3(0.0));
```
7) In `blurAndOutput` change ternary ops to ifs (readability)

**Comment by devshgraphicsprogramming** [comment]
don't forget about trying to compile without the clamp() in load and store Shared

**Comment by devshgraphicsprogramming** [comment]
and if you want to clamp, you have an error
```
    _idx = clamp(_idx, 0u, uint(SMEM_SIZE));
```
should be
```
    _idx = clamp(_idx, 0u, uint(SMEM_SIZE-1));
```

You could try being a sneaky fellow and do
```
    _idx &= uint(SMEM_SIZE-1);
```
and see if the compiler lets you through.

**Comment by devshgraphicsprogramming** [comment]
Good news is that you're not shared memory limited, but I found that for a funny reason the shader works if clamping is enabled and `#define SMEM_SIZE (PS_SIZE+32u)`

P.S. I actually compared a Kogge-Stone Scan (referenced to as Naive Scan by Nvidia) on the 1024 and 512 textures, and it was same speed.

**Comment by devshgraphicsprogramming** [comment]
After doing points 1-7 we can optimize further with things like
http://www.idav.ucdavis.edu/func/return_pdf?pub_id=1041

Looking at `scan_warp`
Essentially you can always assume that threads in consecutive groups of 4 (_tid&4u) run in lock-step so there is no barrier needed to synchronise subgroups of 4 (or when the subgroup extensions are available, the subgroup size).

You could allow the first N (N=2 by default ,or statically queried [4 or 5] with C++ for NV_shader_thread_group or variable for ARB_shader_ballot) invocations of downsweep to not require barriers between them.
Likewise with upsweep, but in the case of last invocations.

The entire algorithm looks like this
1. Scan all warps in parallel using inclusive scan_warp() .
2. Record the last partial result from each warp i
3. With a single warp, perform scan_warp() on the partial results from
Step 2.
4. For each thread of warp i, accumulate the partial results from Step 3
into that thread’s output element from Step 1.

Then use
https://developer.nvidia.com/reading-between-threads-shader-intrinsics
to make the `scan_warp` routine use absolutely no shared memory! (except for the output value of `scan_warp`)
This will reduce shared memory usage by a factor of 4 up to 32 !

Essentially if you can enable the NV_shader_thread_shuffle extension then use that, otherwise try to enable ARB_shader_ballot, else fallback to old method.

this would essentially do a scan/prefix sum on a subgroup first (which is 16,32 or 64 threads wide).

I used these extensions in the SSBO Atomics benchmark.

**Comment by devshgraphicsprogramming** [comment]
Forget about point 6

**Comment by devshgraphicsprogramming** [comment]
> After doing points 1-7 we can optimize further with things like

Regarding that, it would seem that the only acceptable scan input sizes would be:
Default: 4,16,64,256,1024
Nvidia warp16: 16,256, 4096
Nvidia warp32: 32,1024
Variable ARB_shader_ballot: gl_SubGroupSizeARB,gl_SubGroupSizeARB^2,...
*Note*:  gl_SubGroupSizeARB is always within the range [4,64]

calculating the amount of shared memory required will be a pain, so during development until the very last moment after all the optimizations we could set it to be 4096 elements (if your 660 cannot take it then 2048).

**Comment by devshgraphicsprogramming** [comment]
I have made some pseudocode
```
/** Array(s) accessed by getter and setter are required to be padded
with N/2 elements every N elements, where N=W or N=VARIABLE_SCAN_SZ .
GL_WARP_SIZE_NV is a compile time constant, unfortunately gl_SubGroupSizeARB 
is a uniform, not a queriable implementation constant.*/
template<W>
uint padAddressForScan(in uint addr)
{
	return addr+((addr>>1u)&(~(W/2u-1u)));
}


template<setter,getter,W,T=decltype(getter)>
T incl_scan_padded_warp##W(in uint idx)
{
	setter(idx+1,getter(idx)+getter(idx+1));
	...
	...
	...

	return getter(idx);
}

template<setter,getter,W,T=decltype(getter)>
T incl_scan_padded_warp(in uint VARIABLE_SCAN_SZ, in uint idx)
{
	setter(idx+1,getter(idx)+getter(idx+1));
	setter(idx+2,getter(idx)+getter(idx+2));
	switch(VARIABLE_SCAN_SZ)
	{
		case 8u:
			setter(idx+4,getter(idx)+getter(idx+4));
		case 16u:
			setter(idx+8,getter(idx)+getter(idx+8));
		case 32u:
			setter(idx+16,getter(idx)+getter(idx+16));
		case 64u:
			setter(idx+32,getter(idx)+getter(idx+32));
	}

	return getter(idx);
}


template<setter,getter,T=decltype(getter)>
T incl_scan_padded_warpARB(in uint idx)
{
	return incl_scan_padded_warp(gl_SubGroupSizeARB,idx);
}

template<setter,getter,T=decltype(getter)>
T incl_scan_padded_warpNV(in uint idx)
{
	return incl_scan_padded_warp##GL_WARP_SIZE_NV(idx);
}

//! these are fun 
template<setter,T>
T incl_scan_padded_warpNVShuffle(in uint idx, T val)
{
	T tmp = shuffleUpNV(val,1u,GL_WARP_SIZE_NV);
	if (gl_ThreadInWarpNV>=1u)
		val += tmp;
	tmp = shuffleUpNV(val,2u,GL_WARP_SIZE_NV);
	if (gl_ThreadInWarpNV>=2u)
		val += tmp;
	tmp = shuffleUpNV(val,4u,GL_WARP_SIZE_NV);
	if (gl_ThreadInWarpNV>=4u)
		val += tmp;
	tmp = shuffleUpNV(val,8u,GL_WARP_SIZE_NV);
	if (gl_ThreadInWarpNV>=8u)
		val += tmp;
#if GL_WARP_SIZE_NV==32u
	tmp = shuffleUpNV(val,16u,GL_WARP_SIZE_NV);
	if (gl_ThreadInWarpNV>=16u)
		val += tmp;
#endif
	return val;
}


//! these are fun
template<setter,T>
T incl_scan_padded_warpAMDBallot(in uint idx, T val)
{
	return addInvocationsInclusiveScanAMD(val);
}

//! these arae fun
template<setter,T>
T incl_scan_padded_warpAMDBallot(in uint idx, T val)
{
        return subgroupInclusiveAdd(val);
}


#if NV
#define LAST_WARP_CHECK if (gl_ThreadInWarpNV == W-1u) //uint(GL_WARP_SIZE_NV) compile time
#elif ARB
#define LAST_WARP_CHECK if (gl_SubGroupInvocationARB == gl_SubGroupSizeARB-1u)
#else
#define LAST_WARP_CHECK if (gl_LocalInvocationID.x&(W-1u) == W-1u) //W compile time
#endif

//! This is a very crappy example
#define SCAN_FORWARD_IMPL(INCL_SCAN_TO_CALL) \
	BARRIER; \
	totalOutputOffset += padAddressForScan(scanningBatchSize); \
	if (gl_LocalInvocationID.x<scanningBatchSize) \
	{ \
		(LAST_WARP_CHECK) \
			setter(totalOutputOffset+padded_tid_div_w,tmp); \
	} \
	scanningBatchSize = scanningBatchSize>>W_LOG2; \
	BARRIER; \
	if (gl_LocalInvocationID.x<scanningBatchSize) \
		tmp = INCL_SCAN_TO_CALL(totalOutputOffset+padded_tid)

#define SCAN_BACK_IMPL() \
	BARRIER; \
	oldBatchSize = scanningBatchSize; \
	scanningBatchSize = scanningBatchSize<<W_LOG2; \
	if (gl_LocalInvocationID.x<scanningBatchSize) \
		tmp = getter(totalOutputOffset+padded_tid_div_w); \
	BARRIER; \
	totalOutputOffset -= padAddressForScan(oldBatchSize); \
	if (gl_LocalInvocationID.x<scanningBatchSize) \
		setter(totalOutputOffset+padded_tid,getter(totalOutputOffset+padded_tid)+tmp)


// might need a barrier before calling
template<setter,getter,W,T=decltype(getter)>
T incl_scan_padded_group()
{
	const uint padded_tid = padAddressForScan(gl_LocalInvocationID.x);
	const uint padded_tid_div_w = padAddressForScan(gl_LocalInvocationID.x>>W_LOG2);

	uint totalOutputOffset = 0u;
	uint scanningBatchSize = gl_WorkGroupSize.x; //can be compile time constant :D

	// generate hierarchical W-wide scans
	T tmp = incl_scan_...(padded_tid);
	// N-1 times, with N=logW(gl_WorkGroupSize.x) or N=ceil(logW(gl_WorkGroupSize.x))
	SCAN_FORWARD_IMPL(USUAL_PREDICATE,incl_scan_...);
	...
	// last warp-scan could be better sized incl_scan_* function
	SCAN_FORWARD_IMPL(USUAL_PREDICATE,incl_scan_...);

	uint oldBatchSize;
	// N-2 times, with N=logW(gl_WorkGroupSize.x) or N=ceil(logW(gl_WorkGroupSize.x))
	SCAN_BACK_IMPL(); //if you use a custom incl_scan_* on last SCAN_FORWARD_IMPL then you need to match accordingly
	...
	SCAN_BACK_IMPL();
	BARRIER;
	
	//assert(totalOutputOffset==0u);
	tmp = getter(padded_tid_div_w);
	BARRIER;
	return getter(padded_tid)+tmp;
}

//! For a 32-size warp and 1024 sized workgroup it would look like this
template<setter,getter,32u,T=decltype(getter)>
T incl_scan_padded_group()
{
	const uint padded_tid = padAddressForScan(gl_LocalInvocationID.x);
	const uint padded_tid_div_w = padAddressForScan(gl_LocalInvocationID.x>>5u);

	uint totalOutputOffset = 0u;
	uint scanningBatchSize = 1024u; //can be compile time constant :D

	T tmp = incl_scan_...(padded_tid);
	SCAN_FORWARD_IMPL(USUAL_PREDICATE,incl_scan_...);

	uint oldBatchSize;
	BARRIER;
	
	tmp = getter(padded_tid_div_w);
	BARRIER;
	return getter(padded_tid)+tmp;
}
```

This is more or less how the optimizations could look.

For NV_shader_thread_shuffle, AMD_shader_ballot and KHR_shader_subgroup

**Comment by devshgraphicsprogramming** [comment]
Only problem is the behaviour of the shuffle/subgroup operators on inactive lanes :(

**Comment by devshgraphicsprogramming** [comment]
Ok, committed and pushed the warp-scan function generator.

So if you generate your functions with `getWarpInclusiveScanFunctionsPadded(EGCO_ADD, EGT_VEC3, "blur","loadShared","storeShared")` then you will get some useful functions such as:
```
vec3 warp_incl_scan_padded4ublur (in vec3 val, in uint idx)
vec3 warp_incl_scan_padded8ublur (in vec3 val, in uint idx)
vec3 warp_incl_scan_padded16ublur (in vec3 val, in uint idx)
vec3 warp_incl_scan_padded32ublur (in vec3 val, in uint idx)
vec3 warp_incl_scan_padded64ublur (in vec3 val, in uint idx)
```

And if you `#define COMM_OP_IDENTITY vec3(0.0,0.0,0.0)' before including/adding the generated warp-scan header you will get the function `vec3 warp_incl_scan_padded_safeblur(vec3 val, in uint idx, in uint VARIABLE_SCAN_SZ)`


With some ingenuitiy you can implement the CGLSLFunctionGenerator::getBlockInclusiveScanFunctionsPadded method so we can get a function like this:
```
vec3 block_incl_scan_paddedblur (in vec3 val, in uint idx)
```

Hint: use a `switch(WARP_SCAN_SIZE)` to choose which `warp_incl_scan_paddedXublur` to choose and generate the hybrid scans for all warp-sizes (4,8,16,32,64). If you find yourself reducing more than the elements in the block then use warp_incl_scan_padded_safeblur for the top-level reduction.

**Comment by devshgraphicsprogramming** [comment]
Let me treat you to come generated code
```glsl

#version 430 core


#define GL_WARP_SIZE_NV 32u

#extension GL_KHR_shader_subgroup_arithmetic: enable

#ifdef GL_KHR_shader_subgroup_arithmetic
    #define SUBGROUP_SIZE gl_SubgroupSize
#else
    #extension GL_NV_shader_thread_shuffle: enable

    #ifdef GL_NV_shader_thread_shuffle
        #extension GL_NV_shader_thread_group: enable

        #ifdef GL_NV_shader_thread_group
            #define SUBGROUP_SIZE GL_WARP_SIZE_NV
        #else
            #define SUBGROUP_SIZE 32u
        #endif
        #define CONSTANT_SUBGROUP_SIZE

        #define subgroupShuffleUpEMUL(VAL,DELTA) shuffleUpNV((VAL),(DELTA),SUBGROUP_SIZE)
    #else
        #extension GL_AMD_shader_ballot: enable

        #ifdef GL_AMD_shader_ballot
            #define SUBGROUP_SIZE gl_SubGroupSizeARB
        #else
            #extension GL_KHR_shader_subgroup_shuffle_relative: enable
            #extension GL_KHR_shader_subgroup_shuffle: enable

            #if defined(GL_KHR_shader_subgroup_shuffle)||defined(GL_KHR_shader_subgroup_shuffle_relative)
                #define SUBGROUP_SIZE gl_SubgroupSize

                #ifdef GL_KHR_shader_subgroup_shuffle_relative
                    #define subgroupShuffleUpEMUL(VAL,DELTA) subgroupShuffleUp((VAL),(DELTA))
                #else
                    #define subgroupShuffleUpEMUL(VAL,DELTA) subgroupShuffle((VAL),gl_SubgroupInvocationID-(DELTA))
                #endif
            #else
                #extension GL_NV_shader_thread_group: enable

                #ifdef GL_NV_shader_thread_group
                    #define SUBGROUP_SIZE GL_WARP_SIZE_NV
                    #define CONSTANT_SUBGROUP_SIZE
                #else
                    #extension GL_ARB_shader_ballot: enable

                    #ifdef GL_ARB_shader_ballot
                        #define SUBGROUP_SIZE gl_SubGroupSizeARB
                    #else
                        #extension GL_KHR_shader_subgroup_basic: enable

                        #ifdef GL_KHR_shader_subgroup_basic
                            #define SUBGROUP_SIZE gl_SubgroupSize
                        #endif // KHR_shader_subgroup_basic
                    #endif // ARB_shader_ballot
                #endif // NV_shader_thread_group
            #endif // KHR_shader_subgroup_shuffle
        #endif // AMD_shader_ballot
    #endif // NV_shader_thread_shuffle
#endif // KHR_shader_subgroup_arithmetic

#ifdef SUBGROUP_SIZE
    #define PROBABLE_SUBGROUP_SIZE SUBGROUP_SIZE
    #ifdef CONSTANT_SUBGROUP_SIZE
        #define CONSTANT_PROBABLE_SUBGROUP_SIZE
    #endif
#else
    #define PROBABLE_SUBGROUP_SIZE 4u //only size we can guarantee on Intel, unless someone can prove to me that Intel uses SIMD16 mode always for Compute
    #define CONSTANT_PROBABLE_SUBGROUP_SIZE
#endif // SUBGROUP_SIZE
                    
#ifndef _IRR_GENERATED_WARP_PAD_FUNCS_INCLUDED_
#define _IRR_GENERATED_WARP_PAD_FUNCS_INCLUDED_
uint warpPadAddress(in uint addr, in uint width) {return addr+((addr>>1u)&(~(width/2u-1u)));}

uint warpPadAddress4(in uint addr) {return addr+((addr>>1u)&0xfffffffeu);}
uint warpPadAddress8(in uint addr) {return addr+((addr>>1u)&0xfffffffcu);}
uint warpPadAddress16(in uint addr) {return addr+((addr>>1u)&0xfffffff8u);}
uint warpPadAddress32(in uint addr) {return addr+((addr>>1u)&0xfffffff0u);}
uint warpPadAddress64(in uint addr) {return addr+((addr>>1u)&0xffffffe0u);}
uint warpPadAddress128(in uint addr) {return addr+((addr>>1u)&0xffffffc0u);}
uint warpPadAddress256(in uint addr) {return addr+((addr>>1u)&0xffffff80u);}
uint warpPadAddress512(in uint addr) {return addr+((addr>>1u)&0xffffff00u);}
uint warpPadAddress1024(in uint addr) {return addr+((addr>>1u)&0xfffffe00u);}
uint warpPadAddress2048(in uint addr) {return addr+((addr>>1u)&0xfffffc00u);}
uint warpPadAddress4096(in uint addr) {return addr+((addr>>1u)&0xfffff800u);}
#endif //_IRR_GENERATED_WARP_PAD_FUNCS_INCLUDED_
            
#ifndef _IRR_GENERATED_WARP_SCAN_PADDED_blur_FUNCS_INCLUDED_
#define _IRR_GENERATED_WARP_SCAN_PADDED_blur_FUNCS_INCLUDED_

#ifndef _IRR_GENERATED_WARP_PAD_FUNCS_INCLUDED_
#define _IRR_GENERATED_WARP_PAD_FUNCS_INCLUDED_
uint warpPadAddress(in uint addr, in uint width) {return addr+((addr>>1u)&(~(width/2u-1u)));}

uint warpPadAddress4(in uint addr) {return addr+((addr>>1u)&0xfffffffeu);}
uint warpPadAddress8(in uint addr) {return addr+((addr>>1u)&0xfffffffcu);}
uint warpPadAddress16(in uint addr) {return addr+((addr>>1u)&0xfffffff8u);}
uint warpPadAddress32(in uint addr) {return addr+((addr>>1u)&0xfffffff0u);}
uint warpPadAddress64(in uint addr) {return addr+((addr>>1u)&0xffffffe0u);}
uint warpPadAddress128(in uint addr) {return addr+((addr>>1u)&0xffffffc0u);}
uint warpPadAddress256(in uint addr) {return addr+((addr>>1u)&0xffffff80u);}
uint warpPadAddress512(in uint addr) {return addr+((addr>>1u)&0xffffff00u);}
uint warpPadAddress1024(in uint addr) {return addr+((addr>>1u)&0xfffffe00u);}
uint warpPadAddress2048(in uint addr) {return addr+((addr>>1u)&0xfffffc00u);}
uint warpPadAddress4096(in uint addr) {return addr+((addr>>1u)&0xfffff800u);}
#endif //_IRR_GENERATED_WARP_PAD_FUNCS_INCLUDED_
            
#undef COMM_OPInvocationsInclusiveScanAMD
#undef subgroupInclusiveCOMM_OP
#undef COMM_OP
#define COMM_OP(X,Y) ((X)+(Y))
#define subgroupInclusiveCOMM_OP(val) subgroupInclusiveAdd(val)
#define COMM_OPInvocationsInclusiveScanAMD(val) addInvocationsInclusiveScanAMD(val)

#undef TYPE
#define TYPE vec3

#undef WARP_SCAN_SIZE
#define WARP_SCAN_SIZE  PROBABLE_SUBGROUP_SIZE

#undef PASTER2
#undef PASTER3
#define PASTER2(a,b)    a ## b
#define PASTER3(a,b,c)  a ## b ## c
                      
#undef WARP_SCAN_FUNC_NAME
#define WARP_SCAN_FUNC_NAME warp_incl_scan_paddedblur

#undef WARP_SCAN_FUNC_NAME_SAFE
#define WARP_SCAN_FUNC_NAME_SAFE warp_incl_scan_padded_safeblur

#undef WARP_SCAN_FUNC_NAME_SIZED
#define WARP_SCAN_FUNC_NAME_SIZED(W) PASTER3(warp_incl_scan_padded,W,blur)

#undef GETTER
#define GETTER(IDX) loadShared(IDX)

#undef SETTER
#define SETTER(IDX,VAL) storeShared(IDX,VAL)

//! Warning these functions will reduce across the entire warp under GL_KHR_shader_subgroup_arithmetic and GL_AMD_shader_ballot
TYPE WARP_SCAN_FUNC_NAME (in TYPE val, in uint idx);

// base case
TYPE WARP_SCAN_FUNC_NAME_SIZED(4u) (in TYPE val, in uint idx)
{
#if defined(GL_KHR_shader_subgroup_arithmetic)||defined(GL_AMD_shader_ballot)
    return WARP_SCAN_FUNC_NAME (val,idx);
#elif defined(GL_NV_shader_thread_shuffle)||defined(GL_KHR_shader_subgroup_shuffle)||defined(GL_KHR_shader_subgroup_shuffle_relative)
    TYPE tmpA = val;
    TYPE tmpB = subgroupShuffleUpEMUL(tmpA,1u);
    if (gl_ThreadInWarpNV>1u)
        tmpA = COMM_OP(tmpA,tmpB);
    tmpB = subgroupShuffleUpEMUL(tmpA,2u);
    if (gl_ThreadInWarpNV>2u)
        tmpA = COMM_OP(tmpA,tmpB);
    return tmpA;
#elif defined(GL_KHR_shader_subgroup_basic)
    subgroupBarrier();
    SETTER(idx+1u,COMM_OP(GETTER(idx),GETTER(idx+1u)));
    subgroupBarrier();
    SETTER(idx+2u,COMM_OP(GETTER(idx),GETTER(idx+2u)));
    subgroupBarrier();
    return GETTER(idx);
#else
    SETTER(idx+1u,COMM_OP(GETTER(idx),GETTER(idx+1u)));
    SETTER(idx+2u,COMM_OP(GETTER(idx),GETTER(idx+2u)));
    return GETTER(idx);
#endif
}

// recursive
#undef WARP_INCL_SCAN_PADDED_DECL
#define WARP_INCL_SCAN_PADDED_DECL(W)    TYPE WARP_SCAN_FUNC_NAME_SIZED(W) (in TYPE val, in uint idx)

#define GET_LOWER_WARP_SCAN_FUNC_NAME(currentSize) PASTER2(GET_LOWER_WARP_SCAN_FUNC_NAME_,currentSize)
#define GET_LOWER_WARP_SCAN_FUNC_NAME_8u WARP_SCAN_FUNC_NAME_SIZED(4u)
#define GET_LOWER_WARP_SCAN_FUNC_NAME_16u WARP_SCAN_FUNC_NAME_SIZED(8u)
#define GET_LOWER_WARP_SCAN_FUNC_NAME_32u WARP_SCAN_FUNC_NAME_SIZED(16u)
#define GET_LOWER_WARP_SCAN_FUNC_NAME_64u WARP_SCAN_FUNC_NAME_SIZED(32u)

#undef WARP_INCL_SCAN_PADDED_DEF
#if defined(GL_KHR_shader_subgroup_arithmetic)||defined(GL_AMD_shader_ballot)
    #define WARP_INCL_SCAN_PADDED_DEF(W) WARP_INCL_SCAN_PADDED_DECL(W) \
    {\
        return WARP_SCAN_FUNC_NAME (val,idx); \
    }
#elif defined(GL_NV_shader_thread_shuffle)||defined(GL_KHR_shader_subgroup_shuffle)||defined(GL_KHR_shader_subgroup_shuffle_relative)
    #define WARP_INCL_SCAN_PADDED_DEF(W) WARP_INCL_SCAN_PADDED_DECL(W) \
    {\
        TYPE tmpA = GET_LOWER_WARP_SCAN_FUNC_NAME(W) (val,idx); \
        TYPE tmpB = subgroupShuffleUpEMUL(tmpA,W/2u); \
        if (gl_ThreadInWarpNV>(W/2u)) \
            tmpA = COMM_OP(tmpA,tmpB); \
        return tmpA; \
    }
#elif defined(GL_KHR_shader_subgroup_basic)
    #define WARP_INCL_SCAN_PADDED_DEF(W) WARP_INCL_SCAN_PADDED_DECL(W) \
    {\
        SETTER(idx+(W/2u),COMM_OP(GET_LOWER_WARP_SCAN_FUNC_NAME(W) (val,idx),GETTER(idx+(W/2u)))); \
        subgroupBarrier(); \
        return GETTER(idx); \
    }
#else
    #define WARP_INCL_SCAN_PADDED_DEF(W) WARP_INCL_SCAN_PADDED_DECL(W) \
    {\
        SETTER(idx+(W/2u),COMM_OP(GET_LOWER_WARP_SCAN_FUNC_NAME(W) (val,idx),GETTER(idx+(W/2u)))); \
        return GETTER(idx); \
    }
#endif


WARP_INCL_SCAN_PADDED_DEF(8u)

WARP_INCL_SCAN_PADDED_DEF(16u)

WARP_INCL_SCAN_PADDED_DEF(32u)

WARP_INCL_SCAN_PADDED_DEF(64u)


TYPE WARP_SCAN_FUNC_NAME (in TYPE val, in uint idx)
{
#ifdef GL_KHR_shader_subgroup_arithmetic
    return subgroupInclusiveCOMM_OP(val);
#elif defined(GL_NV_shader_thread_shuffle)
    #if SUBGROUP_SIZE==32u
        return WARP_SCAN_FUNC_NAME_SIZED(32u) (val,idx);
    #elif SUBGROUP_SIZE==16u
        return WARP_SCAN_FUNC_NAME_SIZED(16u) (val,idx);
    #else
        #error "What size are your NV warps!?"
    #endif
#elif defined(GL_AMD_shader_ballot)
    return COMM_OPInvocationsInclusiveScanAMD(val);
#elif defined(CONSTANT_PROBABLE_SUBGROUP_SIZE)
    return WARP_SCAN_FUNC_NAME_SIZED(WARP_SCAN_SIZE) (val,idx);
#else
	switch(WARP_SCAN_SIZE)
	{
		case 8u:
			return WARP_SCAN_FUNC_NAME_SIZED(8u) (val,idx);
            break;
		case 16u:
			return WARP_SCAN_FUNC_NAME_SIZED(16u) (val,idx);
            break;
		case 32u:
			return WARP_SCAN_FUNC_NAME_SIZED(32u) (val,idx);
            break;
		case 64u:
			return WARP_SCAN_FUNC_NAME_SIZED(64u) (val,idx);
            break;
	}

    return WARP_SCAN_FUNC_NAME_SIZED(4u) (val,idx);
#endif
}

// COMM_OP_IDENTITY will be user provided
#ifdef COMM_OP_IDENTITY
    //! safe function which will not reduce more of the warp than VARIABLE_SCAN_SZ
    // optimized for a constant/dynamically uniform VARIABLE_SCAN_SZ
    TYPE WARP_SCAN_FUNC_NAME_SAFE (TYPE val, in uint idx,in uint VARIABLE_SCAN_SZ)
    {
        val = idx<VARIABLE_SCAN_SZ ? val:COMM_OP_IDENTITY;
        return WARP_SCAN_FUNC_NAME (val,idx);
    }
#endif
                      
#undef WARP_INCL_SCAN_PADDED_DEF

#undef GET_LOWER_WARP_SCAN_FUNC_NAME
#undef GET_LOWER_WARP_SCAN_FUNC_NAME_8u
#undef GET_LOWER_WARP_SCAN_FUNC_NAME_16u
#undef GET_LOWER_WARP_SCAN_FUNC_NAME_32u
#undef GET_LOWER_WARP_SCAN_FUNC_NAME_64u

#undef WARP_INCL_SCAN_PADDED_DECL
#undef SETTER
#undef GETTER
#undef WARP_SCAN_FUNC_NAME_SIZED
#undef WARP_SCAN_FUNC_NAME_SAFE
#undef WARP_SCAN_FUNC_NAME

#undef PASTER2
#undef PASTER3

#undef TYPE
#undef COMM_OPInvocationsInclusiveScanAMD
#undef subgroupInclusiveCOMM_OP
#undef COMM_OP

#endif // header guard
                      

#define PS_SIZE 1024 // ACTUAL_SIZE padded to PoT
#define ACTUAL_SIZE 1024
#define ACTUAL_SIZE_X 1024
#define ACTUAL_SIZE_Y 1024
#define SSBO_OUT_OFFSET (ACTUAL_SIZE_X*ACTUAL_SIZE_Y) // offset where first pass outputs
#define WG_SIZE 1024 // min(ACTUAL_SIZE, CBlurPerformer::s_MAX_WORK_GROUP_SIZE)
#define SUB_THR_CNT 1 // number of threads simulated by single thread
#define NUM_PASSES 2 // number of passes per axis (i.e. in single shader)
layout(local_size_x = WG_SIZE) in;

#define NUM_BANKS 32
#define LOG_NUM_BANKS 5
#define CONFLICT_FREE_OFFSET(n) ((n) >> LOG_NUM_BANKS)

#define BARRIER \
    memoryBarrierShared();\
    barrier()

layout(std430, binding = 0) restrict buffer Samples {
    uvec2 samples[];
};

#define LC_IDX gl_LocalInvocationIndex
#define FINAL_PASS 0

#if FINAL_PASS
layout(binding = 0, rgba16f) uniform writeonly image2D out_img;
#else
layout(binding = 0) uniform sampler2D in_tex;
#endif


layout(std140, binding = 0) uniform Controls
{
    float u_radius;
};

#define SMEM_SIZE (2*PS_SIZE)
shared float smem_r[SMEM_SIZE];
shared float smem_g[SMEM_SIZE];
shared float smem_b[SMEM_SIZE];


uvec2 encodeRgb(vec3 _rgb)
{
    uvec2 ret;
    ret.x = packHalf2x16(_rgb.rg);
    ret.y = packHalf2x16(vec2(_rgb.b, 0.f));

    return ret;
}
vec3 decodeRgb(uvec2 _rgb)
{
    vec3 ret;
    ret.rg = unpackHalf2x16(_rgb.x);
    ret.b = unpackHalf2x16(_rgb.y).x;

    return ret;
}
 // here goes CS_CONVERSIONS

uint getAddr(uint _addr)
{
    return _addr + CONFLICT_FREE_OFFSET(_addr);
}
void storeShared(uint _idx, vec3 _val)
{
    _idx = clamp(_idx, 0u, uint(SMEM_SIZE-1));
    smem_r[_idx] = _val.r;
    smem_g[_idx] = _val.g;
    smem_b[_idx] = _val.b;
}
vec3 loadShared(uint _idx)
{
    _idx = clamp(_idx, 0u, uint(SMEM_SIZE-1));
    return vec3(smem_r[_idx], smem_g[_idx], smem_b[_idx]);
}

#if FINAL_PASS==0
void loadFromTexture(uint _tid)
{
    if (_tid >= ACTUAL_SIZE)
        return;

    vec2 coords = vec2(float(_tid), float(gl_WorkGroupID.y)) / vec2(ACTUAL_SIZE_X, ACTUAL_SIZE_Y);
    vec4 avg = (
        texture(in_tex, coords) +
        textureOffset(in_tex, coords, ivec2(1, 0)) +
        textureOffset(in_tex, coords, ivec2(0, 1)) +
        textureOffset(in_tex, coords, ivec2(1, 1))
    ) / 4.f;

    storeShared(getAddr(_tid), avg.rgb);
}
#endif

void upsweep(int d, uint offset, uint _tid)
{
    if (_tid < d)
    {
        uint ai = offset*(2*_tid+1)-1;
        uint bi = offset*(2*_tid+2)-1;
        ai += CONFLICT_FREE_OFFSET(ai);
        bi += CONFLICT_FREE_OFFSET(bi);

        storeShared(bi, loadShared(bi) + loadShared(ai));
    }
}
void downsweep(int d, uint offset, uint _tid)
{
    if (_tid < d)
    {
        uint ai = offset*(2*_tid+1)-1;
        uint bi = offset*(2*_tid+2)-1;
        ai += CONFLICT_FREE_OFFSET(ai);
        bi += CONFLICT_FREE_OFFSET(bi);

        vec3 tmp = loadShared(ai);
        storeShared(ai, loadShared(bi));
        storeShared(bi, loadShared(bi) + tmp);
    }
}
void loadFromGmem(uint _inStartIdx, uint _tid)
{
    if (_tid >= ACTUAL_SIZE)
        return;

    const uint ai = _tid;
    const uint bi = _tid + PS_SIZE/2;
    const uint bankOffsetA = CONFLICT_FREE_OFFSET(ai);
    const uint bankOffsetB = CONFLICT_FREE_OFFSET(bi);

    storeShared(ai + bankOffsetA,
	    (ai < ACTUAL_SIZE) ? decodeRgb(samples[_inStartIdx + SSBO_OUT_OFFSET + ai]) : vec3(0)
    );
    storeShared(bi + bankOffsetB,
	    (bi < ACTUAL_SIZE) ? decodeRgb(samples[_inStartIdx + SSBO_OUT_OFFSET + bi]) : vec3(0)
    );
}
void exPsumInSmem()
{
    uint offset = 1;

    BARRIER;
for (uint i = 0; i < 1; ++i) upsweep(512, offset, LC_IDX + i*WG_SIZE);
offset *= 2;
BARRIER;
for (uint i = 0; i < 1; ++i) upsweep(256, offset, LC_IDX + i*WG_SIZE);
offset *= 2;
BARRIER;
for (uint i = 0; i < 1; ++i) upsweep(128, offset, LC_IDX + i*WG_SIZE);
offset *= 2;
BARRIER;
for (uint i = 0; i < 1; ++i) upsweep(64, offset, LC_IDX + i*WG_SIZE);
offset *= 2;
BARRIER;
for (uint i = 0; i < 1; ++i) upsweep(32, offset, LC_IDX + i*WG_SIZE);
offset *= 2;
BARRIER;
for (uint i = 0; i < 1; ++i) upsweep(16, offset, LC_IDX + i*WG_SIZE);
offset *= 2;
BARRIER;
for (uint i = 0; i < 1; ++i) upsweep(8, offset, LC_IDX + i*WG_SIZE);
offset *= 2;
BARRIER;
for (uint i = 0; i < 1; ++i) upsweep(4, offset, LC_IDX + i*WG_SIZE);
offset *= 2;
BARRIER;
for (uint i = 0; i < 1; ++i) upsweep(2, offset, LC_IDX + i*WG_SIZE);
offset *= 2;
BARRIER;
for (uint i = 0; i < 1; ++i) upsweep(1, offset, LC_IDX + i*WG_SIZE);
offset *= 2;
 // here goes unrolled upsweep loop

    if (LC_IDX == 0) { storeShared(PS_SIZE-1 + CONFLICT_FREE_OFFSET(PS_SIZE-1), vec3(0)); }
    memoryBarrierShared();
    barrier();

    BARRIER;
offset /= 2;
for (uint i = 0; i < 1; ++i) downsweep(1, offset, LC_IDX + i*WG_SIZE);
BARRIER;
offset /= 2;
for (uint i = 0; i < 1; ++i) downsweep(2, offset, LC_IDX + i*WG_SIZE);
BARRIER;
offset /= 2;
for (uint i = 0; i < 1; ++i) downsweep(4, offset, LC_IDX + i*WG_SIZE);
BARRIER;
offset /= 2;
for (uint i = 0; i < 1; ++i) downsweep(8, offset, LC_IDX + i*WG_SIZE);
BARRIER;
offset /= 2;
for (uint i = 0; i < 1; ++i) downsweep(16, offset, LC_IDX + i*WG_SIZE);
BARRIER;
offset /= 2;
for (uint i = 0; i < 1; ++i) downsweep(32, offset, LC_IDX + i*WG_SIZE);
BARRIER;
offset /= 2;
for (uint i = 0; i < 1; ++i) downsweep(64, offset, LC_IDX + i*WG_SIZE);
BARRIER;
offset /= 2;
for (uint i = 0; i < 1; ++i) downsweep(128, offset, LC_IDX + i*WG_SIZE);
BARRIER;
offset /= 2;
for (uint i = 0; i < 1; ++i) downsweep(256, offset, LC_IDX + i*WG_SIZE);
BARRIER;
offset /= 2;
for (uint i = 0; i < 1; ++i) downsweep(512, offset, LC_IDX + i*WG_SIZE);
 // here goes unrolled downsweep loop
}

// WARNING: calculates resulting address (this function do NOT expect to get address from getAddr())
vec3 loadSharedInterp(float _idx)
{
    uint floored = uint(_idx);
    return mix(loadShared(getAddr(floored)), loadShared(getAddr(floored+1u)), _idx-float(floored));
}

vec3 blur(float RADIUS, uint _tid)
{
    if (_tid >= ACTUAL_SIZE)
        return vec3(0.f);

    // all index constants below (except LAST_IDX) are enlarged by 1 becaue of **exclusive** prefix sum
    const float FIRST_IDX_F = 1.f;
    const uint FIRST_IDX = 1;
    const uint LAST_IDX = ACTUAL_SIZE-1;
    const float L_IDX = float(_tid) - RADIUS;
    const float R_IDX = float(_tid) + RADIUS + 1.f;
    const float R_EDGE_IDX = min(R_IDX, float(LAST_IDX));

    vec3 res;
    if (R_IDX < float(LAST_IDX))
        res = loadSharedInterp(R_IDX);
    else
        res = (R_IDX+ 1.f - float(LAST_IDX))*loadShared(getAddr(LAST_IDX)) - (R_IDX - float(LAST_IDX))*loadShared(getAddr(LAST_IDX-1));
    if (L_IDX < FIRST_IDX_F)
        res -= (L_IDX - FIRST_IDX_F + 1.f) * loadShared(getAddr(FIRST_IDX));
    else
        res -= loadSharedInterp(L_IDX);

    return res / (2.f*RADIUS + 1.f);
}

void outputFunc(uint _tid)
{
    if (_tid >= ACTUAL_SIZE)
        return;

    uvec2 IDX =
#if FINAL_PASS
    uvec2(gl_WorkGroupID.y, _tid);
#else
    uvec2(_tid, gl_WorkGroupID.y);
#endif

    vec3 res = loadShared(getAddr(_tid));
#if FINAL_PASS
    imageStore(out_img, ivec2(IDX), vec4(res, 1.f));
#else
    samples[IDX.x*ACTUAL_SIZE_Y + IDX.y + SSBO_OUT_OFFSET] = encodeRgb(res);
#endif
}

void main()
{
#if FINAL_PASS
    const uint IN_START_IDX = gl_WorkGroupID.y*ACTUAL_SIZE;
    for (uint i = 0; i < 1; ++i) loadFromGmem(IN_START_IDX, LC_IDX + i*WG_SIZE);
 // here goes loadFromGmem calls
#else
    for (uint i = 0; i < 1; ++i) loadFromTexture(LC_IDX + i*WG_SIZE);
 // here goes loadFromTexture calls
#endif

    const float RADIUS = float(ACTUAL_SIZE) * u_radius;
    for (uint pass = 0u; pass < NUM_PASSES; ++pass)
    {
        exPsumInSmem();

        memoryBarrierShared();
        barrier();

        vec3 blurred[SUB_THR_CNT];
        blurred[0] = blur(RADIUS, LC_IDX + 0*WG_SIZE);
 // here goes blur() calls

        barrier();

        if (LC_IDX + 0*WG_SIZE < ACTUAL_SIZE) storeShared(getAddr(LC_IDX + 0*WG_SIZE), blurred[0]);
 // here goes (local array)->(SMEM) output loop

        memoryBarrierShared();
        barrier();
    }
    for (uint i = 0; i < 1; ++i) outputFunc(LC_IDX + i*WG_SIZE);
 // here goes output loop
}
```

**Comment by devshgraphicsprogramming** [comment]
As a sidenote, your 'CONFLICT_FREE_OFFSET' bank addressing scheme gives a 100% perf increase overall (meaning 300% on just the parallel scan) on my mobile 1060

**Comment by devshgraphicsprogramming** [comment]
I just hit 0.7ms on my 1060 in the latest commit

Here I fixed some things to use `getAddr` 
https://github.com/buildaworldnet/IrrlichtBAW/commit/150e41d2703e8887ea964f39bec124b590041cec

With the implementation of `getNeededSharedElementsForBlockScan` we could get away with declaring an order of magnitude less shared memory as needed when built-in hardware scans or shuffle intrinsics are available (by a factor of 4, 16 or 32)

**Comment by Crisspl** [comment]
O you could merge with my fork before changing blur. There gotta be crazy merging, And some of changes you did was already done on my side

**Comment by devshgraphicsprogramming** [comment]
Ooops, we should really use branches from now on :)

**Comment by Crisspl** [comment]
Whatever, some day we'll have to merge anyway XD

**Comment by devshgraphicsprogramming** [comment]
Since now we only feed our backend SPIR-V shaders we only need to concern ourselves with `GL_KHR_shader_subgroup` and
https://github.com/KhronosGroup/SPIRV-Cross/issues/1350
https://github.com/KhronosGroup/SPIRV-Cross/issues/1351
https://github.com/KhronosGroup/SPIRV-Cross/issues/1352
```glsl
#version 430 core

#ifdef IRR_GL_KHR_shader_subgroup_arithmetic
    #extension GL_KHR_shader_subgroup_arithmetic : enable

    #extension GL_KHR_shader_subgroup_basic: require
    #define SUBGROUP_SIZE gl_SubgroupSize
// do we even need `GL_KHR_shader_subgroup_shuffle_relative` for anything?
#elif defined(IRR_GL_KHR_shader_subgroup_shuffle)||defined(IRR_GL_KHR_shader_subgroup_shuffle_relative)
            #extension GL_KHR_shader_subgroup_shuffle_relative: enable
            #extension GL_KHR_shader_subgroup_shuffle: enable

    #extension GL_KHR_shader_subgroup_basic: require
    #define SUBGROUP_SIZE gl_SubgroupSize
#elif  
    #extension GL_KHR_shader_subgroup_basic: require
    #define SUBGROUP_SIZE gl_SubgroupSize
#endif // KHR_shader_subgroup_shuffle
#endif // KHR_shader_subgroup_arithmetic

// figure out subgroup size and if its constant
#ifdef IRR_GL_IRR_constant_subgroup_size
    #define PROBABLE_SUBGROUP_SIZE IRR_GL_IRR_constant_subgroup_size
    #define CONSTANT_PROBABLE_SUBGROUP_SIZE
#elif defined(SUBGROUP_SIZE)
    #define PROBABLE_SUBGROUP_SIZE SUBGROUP_SIZE
#else
    #define PROBABLE_SUBGROUP_SIZE 4u //only size we can guarantee on Intel, unless someone can prove to me that Intel uses SIMD16 mode always for Compute
    #define CONSTANT_PROBABLE_SUBGROUP_SIZE
#endif 

// will be emitted if irrlicht detects OpenGL
#ifdef IRR_GL_IRR_implicit_subgroup_memory_barrier
    #define subgroupMemoryBarrierShared
    ...
    etc
    ...
#endif 
```

**Comment by devshgraphicsprogramming** [comment]
its a 5 minute task
